app/workers/messaging.php

```diff
@@ -20,59 +33,197 @@
     {
-        return "mails";
+        return "messaging";
     }
 
+
     public function init(): void
     {
-        $dsn = new DSN(App::getEnv('_APP_SMS_PROVIDER'));
-        $user = $dsn->getUser();
-        $secret = $dsn->getPassword();
-
-        $this->sms = match ($dsn->getHost()) {
-            'mock' => new Mock($user, $secret), // used for tests
-            'twilio' => new Twilio($user, $secret),
-            'text-magic' => new TextMagic($user, $secret),
-            'telesign' => new Telesign($user, $secret),
-            'msg91' => new Msg91($user, $secret),
-            'vonage' => new Vonage($user, $secret),
-            default => null
-        };
-
-        $this->from = App::getEnv('_APP_SMS_FROM');
     }
 
     public function run(): void
     {
-        if (empty(App::getEnv('_APP_SMS_PROVIDER'))) {
-            Console::info('Skipped sms processing. No Phone provider has been set.');
-            return;
+        $project = new Document($this->args['project']);
+        $this->dbForProject = $this->getProjectDB($project);
+
+        $message = $this->dbForProject->getDocument('messages', $this->args['messageId']);
+
+        $provider = $this->dbForProject->getDocument('providers', $message->getAttribute('providerId'));
+
+        $this->processMessage($message, $provider);
+    }
+
+    private function processMessage(Document $messageRecord, Document $providerRecord): void
+    {
+        $provider = match ($providerRecord->getAttribute('type')) {
+            'sms' => $this->sms($providerRecord),
+            'push' => $this->push($providerRecord),
+            'email' => $this->email($providerRecord),
+            default => throw new Exception(Exception::PROVIDER_INCORRECT_TYPE)
+        };
+
+        $recipientsId = $messageRecord->getAttribute('to');
+
+        /**
+        * @var Document[] $recipients
+        */
+        $recipients = [];
+
+        $topics = $this->dbForProject->find('topics', [Query::equal('$id', $recipientsId)]);
+        foreach ($topics as $topic) {
+            $recipients = \array_merge($recipients, $topic->getAttribute('targets'));
         }
 
-        if (empty($this->from)) {
-            Console::info('Skipped sms processing. No phone number has been set.');
-            return;
+        $users = $this->dbForProject->find('users', [Query::equal('$id', $recipientsId)]);
+        foreach ($users as $user) {
+            $recipients = \array_merge($recipients, $user->getAttribute('targets'));
         }
 
-        $message = new SMS(
-            to: [$this->args['recipient']],
-            content: $this->args['message'],
-            from: $this->from,
-        );
+        $targets = $this->dbForProject->find('targets', [Query::equal('$id', $recipientsId)]);
+        $recipients = \array_merge($recipients, $targets);
+        $recipients = \array_filter($recipients, fn (Document $recipient) => $recipient->getAttribute('providerId') === $providerRecord->getId());
```

Comment:
Let's line break to make this easier to read:

```suggestion
$recipients = \array_filter($recipients,
    fn (Document $recipient) => $recipient->getAttribute('providerId') === $providerRecord->getId()
 );
```

Reply:
done this